Vision Service: Add Ollama and Data URL support

sgflt opened the pull request:
Refactored code in a way that will support easy addition of another caption providers.

Related Issues:

- #1 
- #4

sgflt committed:
#1 Prepared app for ollama integration

- reduced copy and paste JSON structures
- extracted model management to a separate class

sgflt committed:
#1 Draft of ollama integration

sgflt committed:
#4 Added support for directly uploaded images

sgflt committed:
#1 Pass selected model to ollama dynamically

- Now each implementation of ImageProcessor tests whether it can process a request. The first eligible implementation is used.

sgflt committed:
#1 Dynamically retrieve prompt from environment

- it would be better create configuration file with separate prompts for each model and use case

sgflt committed:
#1 Implemented external AI labeling

sgflt committed:
#1 Improved consistency of output

- if there is label, it must have name and confidence
- we want only short descriptive labels

lastzero commented:
@sgflt Thank you very much! :partying_face:

See comment in the related issue for feedback:
- https://github.com/photoprism/photoprism-vision/issues/1#issuecomment-2850214958

sgflt committed:
#1 Implemented NSFW detection

- local tensorflow model (better performance)
- ollama best effort (may hallucinate)

sgflt committed:
#1 Implemented lazy loading of models and integrations

sgflt committed:
#1 Use proper logging in local_processor

sgflt committed:
#1 Chosen better name for local nsfw model

sgflt committed:
#1 Split single class into class per model

sgflt committed:
#1 Use all available images for label generation

lastzero commented:
@sgflt Awesome, that looks like a huge improvement! :1st_place_medal:  

How about adding some documentation about the new features and changes? :thinking: 

1. It looks like `OLLAMA_ENABLED` needs to be set to `"true"` in the `compose.yaml` configuration and prompts can optionally be customized with `OLLAMA_NSFW_PROMPT`, `OLLAMA_CAPTION_PROMPT` and/or `OLLAMA_LABELS_PROMPT`?
2. The **model name** to pass to the service [probably has to match one of these](https://github.com/ollama/ollama?tab=readme-ov-file#model-library) when Ollama is enabled?
3. Are the **previously implemented models** still available or **do any of the defaults change** when Ollama is enabled?
4. Potential users might also be interested in the **hardware requirements** (depending on the endpoint and/or model) and **which models are recommended**, especially for Ollama?
5. For example, I'm using an **Nvidia RTX 4060 with 8 GB of RAM** in my Linux workstation, which generally allows only small LLMs to be loaded. Would that still be enough? There's a good chance that not all models will fit in memory at the same time, but if the same model can be reused for all endpoints, that's not a big deal.
6. Some users have also asked if it's possible to **run one of these models without a GPU, just on the CPU**? If so, can this be recommended or would it be too slow anyway?

To reduce the effort, it would be sufficient to add this information to the existing [README.md](https://github.com/photoprism/photoprism-vision/blob/develop/README.md).

It also seems like a good idea to include the (most important) variables in the [compose.yaml](https://github.com/photoprism/photoprism-vision/blob/develop/compose.yaml) configuration example, so that users see them even if they don't read the docs. Note that the `service/compose.yaml` file is for development only and not strictly needed when you run the service locally with Python.

Don't worry too much about typos, as we can edit the docs later to improve readability. If you have any questions, please feel free to write an email or get in touch via Element/Matrix! 💬

sgflt commented:
1. added information into README
2. yes, added info into README
3. yes, enabling of Ollama is additive, local configuration will work without any change
4. added some notes
5. dtto 4
6. dtto 4

sgflt committed:
#1 Updated documentation in README.md

lastzero commented:
@sgflt I don't see these changes. Maybe you haven't pushed them yet?

sgflt committed:
#1 Added ollama env variables to compose files

lastzero commented:
Thank you very much! I'll go ahead and merge this if there's nothing you want to add?

sgflt commented:
Okay. I have some minor cleanups as had found that python since 3.9 have better typing hints, but I will do other pull request under different issue.
